packages/server/src/worker/processJobSyncAlgolia.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  JobKind,
+  TJobSyncAlgoliaData,
+  TJobSyncAlgoliaResult,
+} from "./workerTypes";
+import Resource, { resourceDocumentToResource } from "../models/Resource";
+import { TAlgoliaHit, TResource } from "@upswyng/types";
+
+import { Job } from "bullmq";
+import algoliaSearch from "algoliasearch";
+
+/**
+ * Update our Algolia search index with current resource data
+ */
+export async function processJobSyncAlgolia(
+  job: Job<TJobSyncAlgoliaData, TJobSyncAlgoliaResult>,
+  algoliaAppId: string,
+  algoliaIndexName: string,
+  algoliaWriteApiKey: string
+): Promise<TJobSyncAlgoliaResult> {
+  const BATCH_SIZE = 200;
+
+  const estimatedResourceCount = await Resource.estimatedDocumentCount();
+  let batch = 0;
+  let processedResourceCount = 0;
+  let succeeded = false;
+  let deletedRecords: string[] = [];
+  let insertedRecords: string[] = [];
+
+  // Initialize Algolia project and project index
+  const algolia = algoliaSearch(algoliaAppId, algoliaWriteApiKey);
+  const index = algolia.initIndex(algoliaIndexName);
+  console.info(
+    `${job.name}[${job.id}]\t: Algolia connection has been established. Preparing about ${estimatedResourceCount} resources as Algolia Records`
+  );
+
+  // Update (<= BATCH_SIZE) records at a time until resources are exhausted
+  let resources: TResource[] = [];
+  do {
+    console.info(`${job.name}[${job.id}]\tStarting batch ${batch}`);
+    resources = await Promise.all(
+      (
+        await Resource.find({})
+          .skip(batch * BATCH_SIZE)
+          .limit(BATCH_SIZE)
+      ).map(resourceDocumentToResource)
+    );
+    succeeded = false;
+    batch++;
+
+    const updatedAlgoliaIndex = resources.reduce(
+      (
+        container,
+        { resourceId, name, description, subcategories, deleted }
+      ) => {
+        const resourceIndex = {
+          objectID: resourceId,
+          name,
+          description,
+          subcategories: subcategories.map(s => s.name).join(","),
+        };
+
+        deleted
+          ? {
+              toUpdate: container.toUpdate,
+              toDelete: [...container.toDelete, resourceIndex.objectID],
+            }
+          : {
+              toUpdate: [...container.toUpdate, resourceIndex],
+              toDelete: container.toDelete,
+            };
+
+        return container;
+      },
+      {
+        toUpdate: [] as TAlgoliaHit[],
+        toDelete: [] as string[],
+      }
+    );
+
+    console.info(
+      `${job.name}[${job.id}]\t: Algolia Records have been prepared. ${updatedAlgoliaIndex.toUpdate.length} records will be updated and ${updatedAlgoliaIndex.toDelete.length} records will be deleted. Synchronizing Algolia Index...`
+    );
+    try {
+      await Promise.all([
+        index.deleteObjects(updatedAlgoliaIndex.toDelete),
```

Comment:
I think the results of these will give you the no-shit IDs which were updated/deleted. _theoretically_ they should be the same as `toDelete` and `toUpdate` but it's something to consider

Reply:
👀 I like this very much

Reply:
I did not implement this yet because the typing for @types/algoliasearch [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/619f46c319532a2ca9d487723725e76b2f578f33/types/algoliasearch/index.d.ts#L506) does not match the modern api typing [here](https://github.com/algolia/algoliasearch-client-javascript/blob/e6dec843932bdfaac13ffa1e0a11964975499078/packages/client-search/src/methods/index/saveObjects.ts#L18). I think an upgrade to ^4 would be a good idea sometime in the future